rand_core/src/impls.rs

```diff
@@ -56,43 +53,46 @@ pub fn fill_bytes_via_next<R: RngCore + ?Sized>(rng: &mut R, dest: &mut [u8]) {
     }
 }
 
-macro_rules! impl_uint_from_fill {
-    ($rng:expr, $ty:ty, $N:expr) => {{
-        debug_assert!($N == size_of::<$ty>());
-
-        let mut int: $ty = 0;
-        unsafe {
-            let ptr = &mut int as *mut $ty as *mut u8;
-            let slice = slice::from_raw_parts_mut(ptr, $N);
-            $rng.fill_bytes(slice);
-        }
-        int
-    }};
-}
-
 macro_rules! fill_via_chunks {
-    ($src:expr, $dst:expr, $ty:ty, $size:expr) => {{
-        let chunk_size_u8 = min($src.len() * $size, $dst.len());
-        let chunk_size = (chunk_size_u8 + $size - 1) / $size;
+    ($src:expr, $dst:expr, $ty:ty) => {{
+        const SIZE: usize = core::mem::size_of::<$ty>();
+        let chunk_size_u8 = min($src.len() * SIZE, $dst.len());
+        let chunk_size = (chunk_size_u8 + SIZE - 1) / SIZE;
+
         if cfg!(target_endian = "little") {
             unsafe {
-                copy_nonoverlapping(
+                core::ptr::copy_nonoverlapping(
                     $src.as_ptr() as *const u8,
                     $dst.as_mut_ptr(),
                     chunk_size_u8);
             }
         } else {
-            for (&n, chunk) in $src.iter().zip($dst.chunks_mut($size)) {
+            for (&n, chunk) in $src.iter().zip($dst.chunks_mut(SIZE)) {
                 let tmp = n.to_le();
                 let src_ptr = &tmp as *const $ty as *const u8;
                 unsafe {
-                    copy_nonoverlapping(src_ptr,
-                                        chunk.as_mut_ptr(),
-                                        chunk.len());
+                    core::ptr::copy_nonoverlapping(
+                        src_ptr,
+                        chunk.as_mut_ptr(),
+                        chunk.len());
                 }
             }
         }
 
+        // The following code is a safe replacement, but unfortunately ca. 8%
+        // slower.
+        /*
+        let mut iter_src = $src.iter();
+        let mut chunks = $dst.chunks_exact_mut(SIZE);
+        for (chunk, n) in (&mut chunks).zip(&mut iter_src) {
+            chunk.copy_from_slice(&n.to_le_bytes());
+        }
+        let rem = chunks.into_remainder();
+        if let Some(n) = iter_src.next() {
+            rem.copy_from_slice(&n.to_le_bytes()[..rem.len()]);
+        }
+        */
```

Comment:
I don't usually like merging commented code like this — it tends to bit-rot. We *could* make an exception in this case but better just to strip it out of the PR.

Another option would be to use a feature-flag to opt-in to the `unsafe` version, though that doesn't *really* meet the goal of eliminating `unsafe` code. Otherwise I'm still on the fence about switching to this code: 8% slower isn't *too* bad but it's still not ideal.

Reply:
Ok, I'll just remove it.